build-plugins/build-support/src/main/kotlin/aws/sdk/kotlin/gradle/plugins/artifactmetrics/ArtifactMetricsPlugin.kt

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.sdk.kotlin.gradle.plugins.artifactmetrics
+
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.tasks.TaskProvider
+import org.gradle.jvm.tasks.Jar
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.withType
+
+private const val TASK_GROUP = "Verification"
+internal const val OUTPUT_PATH = "reports/metrics/"
+
+/**
+ * Facilitates the collection and analysis of artifact size metrics via the `artifactMetrics` and `analyzeArtifactMetrics` gradle tasks.
+ */
+class ArtifactMetricsPlugin : Plugin<Project> {
+    override fun apply(target: Project) {
+        if (target != target.rootProject) {
+            throw GradleException("${this::class.java} can only be applied to the root project")
+        }
+
+        target.extensions.create("artifactMetricsConfig", ArtifactMetricsConfig::class.java)
+
+        val tasks = mutableListOf<TaskProvider<CollectMetricsTask>>()
+        target.subprojects { tasks.add(subprojectMetricsTask()) }
+
+        target.registerRootProjectMetricsTask(tasks)
+        target.tasks.register<AnalyzeMetricsTask>("analyzeArtifactMetrics") { group = TASK_GROUP }
+    }
+}
+
+private fun Project.subprojectMetricsTask(): TaskProvider<CollectMetricsTask> =
+    tasks.register<CollectMetricsTask>("artifactMetrics") {
+        group = TASK_GROUP
+        onlyIf { tasks.findByName("jvmJar") != null }
```

Comment:
What's the plan for Kotlin/Native artifacts? Maybe add a FIXME here for future support